calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+    ) {
+        for cell in ReadWriteSet::read_set(
+            assignments
+                .iter()
+                .filter(|assign| !Self::reads_only_dones(assign)),
+        ) {
+            if share.is_shareable_component(&cell) {
+                self.reads.insert(cell.borrow().name().clone());
+            }
+        }
+    }
+
+    //Given a control statement c, adds all of the reads of shareable cells from c.
+    //For while loops and if stmts, the control refers only to the guard, not the body.
+    fn get_reads_from_control(
+        &mut self,
+        c: &ir::Control,
+        state_shareable: &ShareSet,
+        is_begin_id: bool,
+    ) {
+        match c {
+            ir::Control::Empty(_)
+            | ir::Control::Seq(_)
+            | ir::Control::Par(_) => {
+                unreachable!("no empty/seqs/pars should be in domination map")
+            }
+            ir::Control::Enable(ir::Enable { group, .. }) => {
+                self.add_assignment_reads(
+                    state_shareable,
+                    &group.borrow().assignments,
+                );
+            }
+            ir::Control::While(ir::While { port, cond, .. }) => {
+                self.add_parent_if_shareable(state_shareable, port);
+                if let Some(group) = cond {
+                    self.add_assignment_reads(
+                        state_shareable,
+                        &group.borrow().assignments,
+                    );
+                }
+            }
+            ir::Control::Invoke(ir::Invoke {
+                comp,
+                inputs,
+                outputs,
+                ..
+            }) => {
+                for (_, port) in inputs.iter() {
+                    self.add_parent_if_shareable(state_shareable, port);
+                }
+                if !outputs.is_empty()
+                    && state_shareable.is_shareable_component(comp)
+                {
+                    self.reads.insert(comp.clone_name());
+                }
+            }
+            ir::Control::If(ir::If { port, cond, .. }) => {
+                if is_begin_id {
+                    self.add_parent_if_shareable(state_shareable, port);
+                    if let Some(group) = cond {
+                        self.add_assignment_reads(
+                            state_shareable,
+                            &group.borrow().assignments,
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+///Contains the name of a cell. The methods in this struct are used to search to
+///see if there was a write to the cell name.
+struct NameSearch {
+    name: ir::Id,
+}
+
+impl NameSearch {
+    fn new(name: ir::Id) -> Self {
+        NameSearch { name }
+    }
+
+    //returns true if the port's parent's name matches name, false otherwise
+    fn parent_matches_name(&self, port: &ir::RRC<ir::Port>) -> bool {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
```

Comment:
You can use the `PortParent::is_hole()` and `PortParent::get_parent_name` methods here instead:
```
let parent = port.borrow().parent;
return !parent.is_hole() && parent.get_parent_name() == self.name
```